### DISCUSSION TITLE
How To Optimize for Dual GPU (4090+1080Ti)?

### DISCUSSION THREAD

**Discussion by YourFriendlyNeighborhoodMONKE** [vladmandic/sdnext#1440] [General]
Couple of people in the 4090 optimization thread of AUTO1111's Discussions pointed out that they got 40it/s speeds out of the box with this repo

This unfortunately didn't happen in my case and I'm seeking help to get my setup optimized

Partially I'm suspecting that my 1080Ti might be interfering inference efficiency somehow
I've tried both SDP and xformers with similar, sub-auto1111 results

Here's the launch dialog:

`15:16:42-939363 INFO     Starting SD.Next
15:16:42-943875 INFO     Python 3.10.11 on Windows
15:16:42-979815 INFO     Version: ba6f9fb4 Thu Jun 15 10:31:30 2023 -0400
15:16:43-224843 INFO     Latest published version: b48a0f13a80cc95b630f0e54e75273d2993a8970 2023-06-15T23:09:29Z
15:16:43-230507 INFO     nVidia CUDA toolkit detected
15:16:44-697189 INFO     Torch 2.0.1+cu118
15:16:44-720274 INFO     Torch backend: nVidia CUDA 11.8 cuDNN 8700
15:16:44-722606 INFO     Torch detected GPU: NVIDIA GeForce RTX 4090 VRAM 24563 Arch (8, 9) Cores 128
15:16:44-724606 INFO     Torch detected GPU: NVIDIA GeForce GTX 1080 Ti VRAM 11264 Arch (6, 1) Cores 28
15:16:44-812917 WARNING  Modified files: ['modules/lora']
15:16:44-850001 INFO     Enabled extensions-builtin: ['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR',
                         'Lora', 'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding',
                         'sd-extension-system-info', 'sd-webui-agent-scheduler', 'sd-webui-controlnet',
                         'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
15:16:44-858508 INFO     Enabled extensions: []
15:16:44-860509 INFO     No changes detected: Quick launch active
15:16:44-866508 INFO     Extension preload: 0.0s M:\SDSYS\automatic\extensions-builtin
15:16:44-868508 INFO     Extension preload: 0.0s M:\SDSYS\automatic\extensions
15:16:44-879789 INFO     Server arguments: []
No module 'xformers'. Proceeding without it.
15:16:51-031227 INFO     Libraries loaded
15:16:51-034737 INFO     Using data path: M:\SDSYS\automatic
15:16:51-093245 INFO     Available VAEs: M:\SDSYS\automatic\models\VAE 1
15:16:51-101250 INFO     Available models: M:\SDSYS\automatic\models\Stable-diffusion 5
15:16:53-504220 INFO     ControlNet v1.1.224
ControlNet preprocessor location: M:\SDSYS\automatic\extensions-builtin\sd-webui-controlnet\annotator\downloads
15:16:53-658904 INFO     ControlNet v1.1.224
Image Browser: ImageReward is not installed, cannot be used.
15:16:54-928121 INFO     Loading UI theme: name=black-orange style=Auto
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
15:16:57-077947 INFO     Local URL: http://127.0.0.1:7860/
15:16:57-078947 INFO     Initializing middleware
15:16:57-207903 INFO     [AgentScheduler] Task queue is empty
15:16:57-210243 INFO     [AgentScheduler] Registering APIs
Loading weights: M:\SDSYS\automatic\models\Stable-diffusion\1536v5e-ep160_+++_1536v2d_ep400__++++__1152_20____++++50-…
15:16:58-645490 INFO     Setting Torch parameters: dtype=torch.float16 vae=torch.float16 unet=torch.float16
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.`

**SDP:**

Here's a screenshot what happens with my GPU's when generating a 512x512 with SDP, sampling set to 99 with Euler_a:

![image](https://github.com/vladmandic/automatic/assets/115096590/1b4379b3-80ca-4283-a06e-5ea7a7301a95)

With 2048x2048 on SDP I'm getting the following:

![image](https://github.com/vladmandic/automatic/assets/115096590/9e9ef3d2-9f38-4720-826c-e6860c598678)

Speeds on the above:
`100%|██████████████████████████████████████████████████████████████████████████████████| 99/99 [00:07<00:00, 14.07it/s]
100%|██████████████████████████████████████████████████████████████████████████████████| 99/99 [02:25<00:00,  1.47s/it]`

Restarting the server with **xformers:**

512x512, Euler_a, 99 sampling:

![image](https://github.com/vladmandic/automatic/assets/115096590/8ce2d31c-8a61-4397-910d-fd5dcb99bc4e)

2048x2048 Euler_a, 99 samples:

![image](https://github.com/vladmandic/automatic/assets/115096590/42a240e1-487c-4acd-bc16-26edad6a6f48)

Midway of the process on 2048x2048 the usage of 1080Ti drops to zero:

![image](https://github.com/vladmandic/automatic/assets/115096590/3b8b80f2-6ec7-4c31-8cb8-c31d22d22ef5)

Inference speeds with xformers on the above:

`100%|██████████████████████████████████████████████████████████████████████████████████| 99/99 [02:05<00:00,  1.26s/it]
100%|██████████████████████████████████████████████████████████████████████████████████| 99/99 [00:06<00:00, 15.30it/s]`

- What do I need to do to optimize the settings to make things work as fast as they should?

**Comment by vladmandic** [comment]
first, tell cuda which device you want to use explicitly:
```batch
set CUDA_VISIBLE_DEVICES="0,1" # run on both
set CUDA_VISIBLE_DEVICES="0" # run on first only
set CUDA_VISIBLE_DEVICES="1" # run on second only
```

second, rtx4090 is never going to get fully saturated on batch size 1 as there is a lot of stop&go because it needs to get info what to do from cpu. and which cpu you have significantly limits how fast 4090 will go.

third, run benchmarks using standardized benchmark tool in sytstem info tab, otherwise your results are non-comparable

**Comment by Tillerz** [comment]
My bet is that the 4090 is so fast compared to the 1080 TI, that it's more work go get it properly running to use both than you actually get a benefit from it. Just a wild guess though.
